### PR TITLE
Logical ai

### PR DESCRIPTION
**Pull request by dylanOmahony2000**
single player ai finished

### CODE REVIEW

**Review by Konrad-Dagiel — CHANGES_REQUESTED** [review]
try to include the character select changes please

File: game/fighter.py [hunk 1]
```diff
@@ -84,15 +93,17 @@ def resource_path(relative_path):
             base_path = os.path.abspath(".")
 
         return os.path.join(base_path, relative_path)
-
+    
     def move(self, screen_width, screen_height, surface, target, obstacles):
-        GRAVITY = 1
+        GRAVITY = 2
```
Comment (KeithS14) [inline game/fighter.py:98]: players cant jump as high now and cant jump to high grounds anymore with gravity = 2

**Comment by dylanOmahony2000** [comment]
> try to include the character select changes please

added in the changes for the character select menu there

**Review by KeithS14 — CHANGES_REQUESTED** [review]
fix players jumping height and comment out line in obstacles.py to hide the obstacles

File: game/fighter.py [hunk 2]
```diff
@@ -34,6 +35,14 @@ def __init__(self, player, x, y, flip, punch_sound, projectile_sound, hit_sound,
...
         self._start_y = y
         self.jump_height = 20
+        self.player_x = x
+        self.player_y = y
+        self.ai = ai
+        self.moves = [0,0,0,0,0,0]
+
+    def return_state(self):
+        return  [self.player_x,self.player_y,self.vel_x,self.vel_y, self.shooting_projectile, self.attacking,
+                        self.running , self.jump , self.blocking]
```
Comment (michaelc445) [inline game/fighter.py:45]: this ai stuff should not be in this fighter class, your fighter class should inherit from this fighter class and add on extra features

File: game/fighter.py [hunk 3]
```diff
@@ -84,15 +93,17 @@ def resource_path(relative_path):
...
         self.dx = 0
         self.dy = 0
 
         # check player 1 movement
-
-        self.keybinds(self.controls, surface, target,None)
+        if self.ai:  # if ai instance call fighter ai func
+            self.keyAi( surface, target,None)
+        else:
+            self.keybinds(self.controls, surface, target,None)
```
Comment (michaelc445) [inline game/fighter.py:106]: same here you should add all this to your ai fighter class in some move_ai function

File: main.py [hunk 4]
```diff
@@ -130,22 +132,35 @@ def game_loop():
...
+        fighter_1 = createNomad(Fighter, 1, p1_spawn[0], p1_spawn[1], False, punch_fx, projectile_fx, hit_fx, player1_controls,False)
 
     elif p1 == "warrior":
-        fighter_1 = createWarrior(Fighter, 1, p1_spawn[0], p1_spawn[1], False, punch_fx, projectile_fx, hit_fx, player1_controls)
+        fighter_1 = createWarrior(Fighter, 1, p1_spawn[0], p1_spawn[1], False, punch_fx, projectile_fx, hit_fx, player1_controls,False)
+
+    if mode == "Single Player":
+        if p2 == "wizard":
+            fighter_2 = createWizard(Fighter_ai, 2, p2_spawn[0], p2_spawn[1], True, punch_fx, projectile_fx, hit_fx,
+                                     player2_controls,True)
```
Comment (michaelc445) [inline main.py:146]: there needs to be a seperate function for singleplayer vs ai and local multiplayer

File: main.py [hunk 5]
```diff
@@ -1336,7 +1361,8 @@ def map_select():
                     mixer.music.load(resource_path("game/assets/audio/background-game.wav"))
                     mixer.music.play(-1)
                     mixer.music.set_volume(background_music_volume)
-                    game_loop()
+                    game_loop(mode)
+
```
Comment (michaelc445) [inline main.py:1365]: make a separate function for this

Comment (michaelc445) [reply main.py:1365]: instead of mode

**Review by michaelc445 — CHANGES_REQUESTED** [review]

File: game/fighter.py (hunk 2 above)
Comment (dylanOmahony2000) [reply game/fighter.py:45]: the return state is needed for the ai to know the fighters position and current action

File: game/fighter.py (hunk 3 above)
Comment (dylanOmahony2000) [reply game/fighter.py:106]: i fixed this problem created a move function in fighter_ai

File: main.py (hunk 4 above)
Comment (dylanOmahony2000) [reply main.py]: I created a single_game_loop function and i returned the gtame_loop function to normal

File: game/fighter_ai.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                self.dx = self.speed
+
+                self.running = True
+
+                self.flip = False
+
+            # jump
+            if self.moves[2]==1 and not self.jump:
+                self.vel_y = -30
+                self.jump = True
```
Comment (michaelc445) [inline game/fighter_ai.py:135]: think this is why your ai is flying

**Review by michaelc445 — CHANGES_REQUESTED** [review]

**Review by michaelc445 — APPROVED** [review]

**Review by KeithS14 — APPROVED** [review]

**Pull request merged**